kernel/src/mem.rs

```diff
@@ -69,213 +82,199 @@ unsafe fn raw_appslice_to_slice<'a>(ptr: *const u8, len: usize) -> &'a [u8] {
     /// reclaimed, this method must return the default value.
     ///
-    /// # Default AppSlice
+    /// # Default Process Buffer
     ///
-    /// A default instance of an AppSlice must return the passed
-    /// default value without executing the closure.
+    /// A default instance of a process buffer must return the passed
+    /// default value without executing the passed closure.
     fn map_or<F, R>(&self, default: R, fun: F) -> R
     where
-        F: FnOnce(&[u8]) -> R;
+        F: FnOnce(&ReadableProcessSlice) -> R;
 }
 
-/// A readable and writable region of userspace memory.
+/// A readable and writeable region of userspace process memory.
 ///
 /// This trait can be used to gain read-write access to memory regions
-/// wrapped in an AppSlice type.
+/// wrapped in a [`ReadWriteProcessBuf`].
 ///
-/// This is a supertrait of [`Read`], which features further methods.
-pub trait ReadWrite: Read {
-    /// Applies a function to the mutable slice reference pointed to
-    /// by the AppSlice.
+/// This is a supertrait of [`ReadableProcessBuf`], which features
+/// methods allowing mutable access.
+pub trait WriteableProcessBuf: ReadableProcessBuf {
+    /// Applies a function to the mutable process slice reference
+    /// pointed to by the [`ReadWriteProcessBuf`].
     ///
     /// If the process is no longer alive and the memory has been
     /// reclaimed, this method must return the default value.
     ///
-    /// # Default AppSlice
+    /// # Default Process Buffer
     ///
-    /// A default instance of an AppSlice must return the passed
+    /// A default instance of a process buffer must return the passed
     /// default value without executing the closure.
-    fn mut_map_or<F, R>(&mut self, default: R, fun: F) -> R
+    fn mut_map_or<F, R>(&self, default: R, fun: F) -> R
     where
-        F: FnOnce(&mut [u8]) -> R;
+        F: FnOnce(&WriteableProcessSlice) -> R;
 }
 
-/// Read-writable memory region of a process, shared with the kernel
-pub struct ReadWriteAppSlice {
-    ptr: *mut u8,
+/// Read-only buffer shared by a userspace process
+///
+/// This struct is provided to capsules when a process Allows a
+/// particular section of its memory to the kernel and gives the
+/// kernel read access to this memory.
+///
+/// It can be used to obtain a [`ReadableProcessSlice`], which is
+/// based around a slice of [`Cell`]s. This is because a userspace can
+/// Allow overlapping sections of memory into different
+/// [`ReadableProcessSlice`]. Having at least one mutable Rust slice
+/// along with read-only slices to overlapping memory in Rust violates
+/// Rust's aliasing rules. A slice of [`Cell`]s avoids this issue by
+/// explicitly supporting interior mutability. Still, a memory barrier
+/// prior to switching to userspace is required, as the compiler is
+/// free to reorder reads and writes, even through [`Cell`]s.
+pub struct ReadOnlyProcessBuf {
+    ptr: *const u8,
```

Comment:
Should `ptr` be a `core::ptr::NonNull` so that `Option<ReadOnlyProcessBuf>` benefits from niche optimizations? That would require changing the `new_external` safety invariants somewhat, but I think it would shrink grant sizes significantly.

The same applies to `ReadWriteProcessBuf`.

Reply:
I suppose the issue with this is that the process is free to Allow a buffer with a null-pointer, so long as the name is zero. Now, that is never valid as a Rust slice reference, which is why were using `NonNull::dangling()` in this case, but we're going to need to provide the process' Allow data back unmodified, which is why presumably we can't use `NonNull` here.

However, `struct ProcessId` starts with a `&'static Kernel`, which is guaranteed to never be null. By making this the first field, we should still benefit from these optimizations, right?

Reply:
Oh, right, `ReadOnlyProcessBuf`/`ReadWriteProcessBuf` are used to return the correct slice back to userspace, not just to access it from capsules.